op-e2e/actions/interop/interop_fork_test.go

```diff
@@ -0,0 +1,325 @@
+	dsl.RequireUnsafeTimeOffset(t, system.Actors.ChainA, 8)
+	dsl.RequireUnsafeTimeOffset(t, system.Actors.ChainB, 8)
+
+	// Submit batch data for each chain in separate L1 blocks so tests can have one chain safe and one unsafe
+	system.SubmitBatchData(func(opts *dsl.SubmitBatchDataOpts) {
+		opts.SetChains(system.Actors.ChainA)
+	})
+	dsl.RequireL1Heads(t, system, 5, 3)
+
+	system.SubmitBatchData(func(opts *dsl.SubmitBatchDataOpts) {
+		opts.SetChains(system.Actors.ChainB)
+	})
+	dsl.RequireL1Heads(t, system, 6, 3)
+
+	// Verify proofs agree on prev blocks
+	assertProgramOutputMatchesDerivationForBlockTimestamp(gt, system, system.Actors.ChainA.Sequencer.L2Safe().Time)
+
+	// Send a message
+	system.AddL2Block(system.Actors.ChainA, dsl.WithL2BlockTransactions(emitter.EmitMessage(alice, "hello")))
+	initMsg := emitter.LastEmittedMessage()
+	system.AddL2Block(system.Actors.ChainB, dsl.WithL2BlockTransactions(system.InboxContract.Execute(alice, initMsg)))
+
+	// Submit batch data for each chain in separate L1 blocks so tests can have one chain safe and one unsafe
+	system.SubmitBatchData(func(opts *dsl.SubmitBatchDataOpts) {
+		opts.SetChains(system.Actors.ChainA)
+	})
+	dsl.RequireL1Heads(t, system, 7, 3)
+	system.SubmitBatchData(func(opts *dsl.SubmitBatchDataOpts) {
+		opts.SetChains(system.Actors.ChainB)
+	})
+	dsl.RequireL1Heads(t, system, 8, 3)
+
+	// Verify chain time
+	dsl.RequireUnsafeTimeOffset(t, system.Actors.ChainA, 10)
+	dsl.RequireUnsafeTimeOffset(t, system.Actors.ChainB, 10)
+
+	// Verify proofs agree on prev blocks
+	assertProgramOutputMatchesDerivationForBlockTimestamp(gt, system, system.Actors.ChainA.Sequencer.L2Safe().Time)
+}
+
+func VerifyContractsDeployedCorrectly(t helpers.Testing, chain *dsl.Chain, activationBlockTxs []*types.Transaction, activationBlockID eth.BlockID) {
+	require.Len(t, activationBlockTxs, 5) // 4 upgrade txs + 1 system deposit tx
+	upgradeTransactions := activationBlockTxs[1:]
+	upgradeTransactionBytes := make([]hexutil.Bytes, len(upgradeTransactions))
+	for i, tx := range upgradeTransactions {
+		txBytes, err := tx.MarshalBinary()
+		require.NoError(t, err)
+		upgradeTransactionBytes[i] = txBytes
+	}
+
+	expectedUpgradeTransactions, err := derive.InteropNetworkUpgradeTransactions()
+	require.NoError(t, err)
+
+	require.Equal(t, upgradeTransactionBytes, expectedUpgradeTransactions)
+
+	RequireContractDeployedAndProxyUpdated(t, chain, derive.CrossL2InboxAddress, predeploys.CrossL2InboxAddr, new(big.Int).SetUint64(activationBlockID.Number))
+	RequireContractDeployedAndProxyUpdated(t, chain, derive.L2ToL2MessengerAddress, predeploys.L2toL2CrossDomainMessengerAddr, new(big.Int).SetUint64(activationBlockID.Number))
+}
+
+// Runs superchain fault proofs across entire superchain pseudo-blocks
+func assertProgramOutputMatchesDerivationForBlockTimestamp(t *testing.T, system *dsl.InteropDSL, endTimestamp uint64) {
+	// Verify the proofs backend agrees
+	actors := system.Actors
+
+	startTimestamp := endTimestamp - 1
```

Comment:
```suggestion
	startTimestamp := endTimestamp - 2
```
since we are testing with two second block time by default?

Reply:
Superchain pseudo-blocks are every one second, and the proof system can only be run on individual pseudo-blocks.

Reply:
See https://specs.optimism.io/interop/fault-proof.html#consolidation for more info.